examples/mnist-classification/src/mnist_classification/core.clj

```diff
@@ -0,0 +1,214 @@
+;; files. The top level folders are named 'training' and 'test'. The subfolders
+;; are named with class names, and those folders are filled with images of the
+;; appropriate class.
+(defn- ds-data->png
+  "Given data from the original dataset, use think.image to produce png data."
+  [ds-data]
+  (let [data-bytes (byte-array (* image-size image-size))
+        num-pixels (alength data-bytes)
+        retval (image/new-image image/*default-image-impl*
+                                image-size image-size :gray)]
+    (c-for [idx 0 (< idx num-pixels) (inc idx)]
+           (let [[x y] [(mod idx image-size)
+                        (quot idx image-size)]]
+             (aset data-bytes idx
+                   (unchecked-byte (* 255.0
+                                      (+ 0.5 (m/mget ds-data y x)))))))
+    (image/array-> retval data-bytes)))
+
+
+(defn- save-image!
+  "Save a dataset image to disk."
+  [output-dir [idx {:keys [data label]}]]
+  (let [image-path (format "%s/%s/%s.png" output-dir (util/max-index label) idx)]
+    (when-not (.exists (io/file image-path))
+      (io/make-parents image-path)
+      (i/save (ds-data->png data) image-path))
+    nil))
+
+;; These two defonces use helpers from cortex to procure the original dataset.
+(defn- timed-get-dataset
+  [f name]
+  (println "Loading" name "dataset.")
+  (let [start-time (System/currentTimeMillis)
+        ds (f)]
+    (println (format "Done loading %s dataset in %ss"
+                     name (/ (- (System/currentTimeMillis) start-time) 1000.0)))
+    ds))
+
+(defonce training-dataset
+  (timed-get-dataset mnist/training-dataset "mnist training"))
+
+(defonce test-dataset
+  (timed-get-dataset mnist/test-dataset "mnist test"))
+
+
+(def dataset-folder "mnist/")
+
+
+(defonce ensure-images-on-disk!
+  (memoize
+   (fn []
+     (println "Ensuring image data is built, and available on disk.")
+     (dorun (map (partial save-image! (str dataset-folder "training"))
+                 (map-indexed vector training-dataset)))
+     (dorun (map (partial save-image! (str dataset-folder "test"))
+                 (map-indexed vector test-dataset)))
+     :done)))
+
+
+(def max-image-rotation-degrees 25)
+
+(defn- image-aug-pipeline
+  "Uses think.image augmentation to vary training inputs."
+  [image]
+  (-> image
+      (image-aug/rotate (- (rand-int (* 2 max-image-rotation-degrees))
+                           max-image-rotation-degrees)
+                        false)
+      (image-aug/inject-noise (* 0.25 (rand)))))
+
+
+
+
+(defn- mnist-observation->image
+  "Creates a BufferedImage suitable for web display from the raw data
+  that the net expects."
+  [observation]
+  (patch/patch->image observation image-size))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The classification experiment system needs a way to go back and forth from
+;; softmax indexes to string class names.
+(def class-mapping
```

Comment:
I think it might be easier to specify these two separately, especially since they could be specified as functions instead of entries in a map. Just a thought though.